streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableKTableJoinMerger.java

```diff
@@ -28,22 +28,52 @@
     private final String queryableName;
     private boolean sendOldValues = false;
 
-    public KTableKTableJoinMerger(final KTableImpl<K, ?, V> parent1,
-                                  final KTableImpl<K, ?, V> parent2,
-                                  final String queryableName) {
+    KTableKTableJoinMerger(final KTableImpl<K, ?, V> parent1,
+                           final KTableImpl<K, ?, V> parent2,
+                           final String queryableName) {
         this.parent1 = parent1;
         this.parent2 = parent2;
         this.queryableName = queryableName;
     }
 
     @Override
     public Processor<K, Change<V>> get() {
-        return new KTableKTableJoinMergeProcessor<>();
+        return new KTableKTableJoinMergeProcessor();
     }
 
     @Override
     public KTableValueGetterSupplier<K, V> view() {
-        return parent1.valueGetterSupplier();
+        // if the result KTable is materialized, use the materialized store to return getter value;
+        // otherwise rely on the parent getter and apply join on-the-fly
+        if (queryableName != null) {
+            return new KTableMaterializedValueGetterSupplier<K, V>(queryableName);
+        } else {
+            return new KTableValueGetterSupplier<K, V>() {
+
+                public KTableValueGetter<K, V> get() {
+                    return parent1.valueGetterSupplier().get();
+                }
+
+                @Override
+                public String[] storeNames() {
+                    // we need to allow the downstream processor to be able to access both ends of the joining table's value getters
+                    final String[] storeNames1 = parent1.valueGetterSupplier().storeNames();
```

Comment:
Why not something like:
```
final List<String> storeNames = Arrays.asList(parent1.valueGetterSupplier().storeNames());
                    storeNames.addAll(Arrays.asList(parent2.valueGetterSupplier().storeNames()));
                    return storeNames.toArray(new String[storeNames.size()]);
```
?
I don't think it is on the critical path so performance shouldn't be an issue

Reply:
What you suggested is actually the original code, while @mjsax suggested doing this. I am actually fine either way.

Reply:
Damian suggest is actually different as it avoid the loops... That would also be fine. But using ArrayList plus for-loops is a mess IMHO.